tests/test-awful-rules.lua

```diff
@@ -282,10 +282,38 @@ test_rule {
 
         assert(c.screen.selected_tag.name ~= "8")
 
+        -- Test the custom sources
+        assert(c.random1)
+        assert(c.random2)
+        assert(not c.random3)
+
         return true
     end
 }
 
+-- Test the custom sources
+assert(awful.rules.add_rule_source("before", function(c, props, _)
+    assert(type(c) == "client")
+    assert(not props.random1)
+
+    props.random2 = true
+end, {"awful.rules"}, {"awful.spawn"}))
```

Comment:
Now that I see this: How about changing the argument order of `add_rule_source` so that the callback is last?